### PR TITLE
Migrated a subset of Xtend code from the "org.eclipse.xtext.ui.testing" project to Java

### PR DESCRIPTION
**Pull request by nbhusare**
- [x] Migrated subset of the Xtend code to Java
- [x] Ran the Maven/Gradle build locally

Signed-off-by: nbhusare <neerajbhusare@gmail.com>

### CODE REVIEW

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractContentAssistTest.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	public XtextResource getResourceFor(InputStream stream) {
+		XtextResourceSet resourceSet = resourceSetProvider.get();
+		initializeTypeProvider(resourceSet);
+		try {
+			URI resourceUri = URI.createURI("Test." + fileExtensionProvider.getPrimaryFileExtension());
+			Resource resource = resourceSet.createResource(resourceUri);
+			resource.load(stream, null);
+			return (XtextResource) resource;
+		} catch (IOException e) {
+			throw new RuntimeException(e);
```
Comment (nbhusare) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractContentAssistTest.java:79]: Altered the code to throw `RuntimeException`.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractContentAssistTest.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+	@BeforeClass
+	@BeforeAll
+	public static void setUp() throws CoreException {
+		javaProject = JavaProjectSetupUtil.createJavaProject("contentAssistTest");
+	}
+
+	@AfterClass
+	@AfterAll
+	public static void tearDown() throws CoreException {
```
Comment (nbhusare) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractContentAssistTest.java:64]: This method now throws a `CoreException`

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractContentAssistTest.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	private FileExtensionProvider fileExtensionProvider;
+
+	@Inject
+	private Injector injector;
+
+	private static IJavaProject javaProject;
+
+	@BeforeClass
+	@BeforeAll
+	public static void setUp() throws CoreException {
```
Comment (nbhusare) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractContentAssistTest.java:58]: This method now throws a `CoreException`

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractEditorDoubleClickTextSelectionTest.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	 *            The text that is expected to be selected after double clicking in the Xtext editor on the current cursor position.
+	 */
+	public void assertSelectedTextAfterDoubleClicking(CharSequence charSequence, String expected) {
+		IFile file = createFile(getContent(charSequence));
+		try {
+			XtextEditor xtextEditor = openEditor(file);
+			xtextEditor = doubleClick(xtextEditor, getCursorPosition(charSequence));
+			assertSelectedText(xtextEditor, expected);
+		} catch (Exception e) {
+			fail("Error opening editor associated with the file " + file.getName());
```
Comment (nbhusare) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractEditorDoubleClickTextSelectionTest.java:59]: Added this line to ensure that the test fails if an exception is thrown, and a proper error message is provided to the user.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractEditorDoubleClickTextSelectionTest.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+	protected String getContent(CharSequence text) {
+		return text.toString().replace(c(), "");
+	}
+
+	protected IFile createFile(String content) {
+		try {
+			return IResourcesSetupUtil.createFile(getProjectName(), getFileName(), getFileExtension(), content);
+		} catch (InvocationTargetException | CoreException | InterruptedException e) {
+			throw new RuntimeException(e);
```
Comment (nbhusare) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractEditorDoubleClickTextSelectionTest.java:77]: Throwing a `RuntimeException` here.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+	public void testFoldingRegions(CharSequence charSequence) {
+		foldingRegionsArePresent(openInEditor(dslFile(charSequence)), charSequence);
+	}
+
+	protected IFile dslFile(CharSequence text) {
+		try {
+			IFile file = IResourcesSetupUtil.createFile(getProjectName(), getFileName(), getFileExtension(), getContent(text));
+
+			/*
+			 * TODO: find a better (with good performance) solution
```
Comment (nbhusare) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java:73]: The **TODO** was already there. Please suggest if there is a better way to deal with this situation. I'll make the required change.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+			 * to set the Xtext nature on the test project.
+			 */
+			IProject project = file.getProject();
+			if (!project.hasNature(XtextProjectHelper.NATURE_ID)) {
+				IResourcesSetupUtil.addNature(project, XtextProjectHelper.NATURE_ID);
+			}
+			return file;
+		} catch (InvocationTargetException | CoreException | InterruptedException e) {
+			throw new RuntimeException(e);
```
Comment (nbhusare) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java:82]: Throwing a `RuntimeException` here.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+		return text.toString().replaceAll(Pattern.quote(b()), "").replaceAll(Pattern.quote(e()), "");
+	}
+
+	protected IXtextDocument openInEditor(IFile dslFile) {
+		try {
+			IXtextDocument document = openEditor(dslFile).getDocument();
+			assertNotNull(document);
+			return document;
+		} catch (Exception e) {
+			throw new RuntimeException(e);
```
Comment (nbhusare) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java:108]: Throwing a `RuntimeException` here.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+				// Insert the beginning folding region marker at the beginning of the corresponding line
+				int startLine = document.getLineOfOffset(foldingRegion.offset);
+				lines.set(startLine, b() + lines.get(startLine));
+
+				// Insert the end folding region marker at the end of the corresponding line
+				int endLine = document.getLineOfOffset(foldingRegion.offset + foldingRegion.length) - 1;
+				lines.set(endLine, lines.get(endLine) + e());
+			}
+		} catch (BadLocationException e) {
+			//Log an exception here???
```
Comment (nbhusare) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java:130]: I believe we should log a message here. @szarnekow what do you think?

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractHoverTest.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+	protected IFile dslFile(CharSequence content) {
+		try {
+			IFile file = IResourcesSetupUtil.createFile(getProjectName(), getFileName(), getFileExtension(), content.toString());
+			IProject project = file.getProject();
+			if (!project.hasNature(XtextProjectHelper.NATURE_ID)) {
+				IResourcesSetupUtil.addNature(project, XtextProjectHelper.NATURE_ID);
+			}
+			return file;
+		} catch (InvocationTargetException | CoreException | InterruptedException e) {
+			throw new RuntimeException(e);
```
Comment (nbhusare) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractHoverTest.java:86]: Throwing a `RuntimeException` here.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractHoverTest.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+		XtextEditor editor = openInEditor(dslFile);
+		Object hoverInfo = ((ITextHoverExtension2) eObjectHover).getHoverInfo2(editor.getInternalSourceViewer(), hoverRegion);
+		return (BrowserInformationControlInput) hoverInfo;
+	}
+
+	protected void hoverPopupHasContent(BrowserInformationControlInput info, String expectedHoverContent) {
+		assertNotNull("No hover found!", info);
+
+		String actualHoverContent = info.getHtml();
+		StringConcatenation builder = new StringConcatenation();
```
Comment (nbhusare) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractHoverTest.java:121]: I am not sure how I could replace the call to `#newLineIfNotEmpty()`, which is why I have kept this code as it is.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractHoverTest.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+	protected String getFileExtension() {
+		return fileExtensionProvider.getPrimaryFileExtension();
+	}
+
+	protected XtextEditor openInEditor(IFile dslFile) {
+		try {
+			syncUtil.waitForBuild(new NullProgressMonitor());
+			return openEditor(dslFile);
+		} catch (Exception e) {
+			throw new RuntimeException(e);
```
Comment (nbhusare) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractHoverTest.java:107]: Throwing a `RuntimeException` here.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/util/AnnotatedTextToString.java [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+				return contents;
+			}
+
+			if (file != null) {
+				return IResourcesSetupUtil.fileToString(file);
+			}
+
+			return null;
+		} catch (CoreException | IOException e) {
+			throw new RuntimeException(e);
```
Comment (nbhusare) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/util/AnnotatedTextToString.java:67]: Throwing a `RuntimeException` here.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/util/AnnotatedTextToString.java [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+					transformed.add(attribute + "=" + marker.getAttribute(attribute));
+				}
+
+				String comment = transformed.stream().collect(Collectors.joining(", "));
+				Integer offset = (Integer) marker.getAttribute(IMarker.CHAR_START);
+				Integer length = (Integer) marker.getAttribute(IMarker.CHAR_END);
+				regions.add(new CommentedRegion(offset, length, comment));
+			}
+		} catch (CoreException e) {
+			// Log the exception?
```
Comment (nbhusare) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/util/AnnotatedTextToString.java:114]: I believe we should log a message here. @szarnekow what do you think?

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/util/AnnotatedTextToString.java [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+		}
+		emptyMessages.add("(no markers found)");
+		return this;
+	}
+
+	public AnnotatedTextToString withMarkersFromFile(String type, String... attributes) {
+		try {
+			return withMarkers(file.findMarkers(type, true, IResource.DEPTH_INFINITE), attributes);
+		} catch (CoreException e) {
+			// Log the exception and return null or throw RuntimeException?
```
Comment (nbhusare) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/util/AnnotatedTextToString.java:124]: I believe we should log a message here, and return a `null` value. @szarnekow what do you think?

**Comment by nbhusare** [comment]
@szarnekow @tivervac @cdietrich Kindly review the PR. I have added notes to indicated some exception handling code that I have added. In few places, I was not sure whether to log a message or throw an exception. Kindly have a look at it and please provide your feedback.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/util/AnnotatedTextToString.java [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+			Integer offset = location.getKey();
+			String comment = location.getValue();
+			result.append(contents.substring(lastOffset, offset));
+			result.append(comment);
+			lastOffset = offset;
+		}
+
+		result.append(contents.substring(lastOffset, contents.length()));
+		String[] resultsArray = result.toString().replace("\t", "    ").split("\r?\n");
+		int maxLineLenght = Arrays.stream(resultsArray).map(r -> r.length()).reduce(Integer::max).get();
```
Comment (szarnekow) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/util/AnnotatedTextToString.java:163]: Typo: length

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractContentAssistTest.java [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+		}
+	}
+
+	@Override
+	public IJavaProject getJavaProject(ResourceSet resourceSet) {
+		return javaProject;
+	}
+
+	protected ArrayList<String> expect(String[]... arrays) {
+		ArrayList<String> expectation = new ArrayList<>();
```
Comment (tivervac) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractContentAssistTest.java:89]: Use List<String> on the left hand side

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/util/AnnotatedTextToString.java [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+
+	private IFile file;
+
+	private List<CommentedRegion> regions = new ArrayList<>();
+
+	private List<String> emptyMessages = new ArrayList<>();
+
+	public String getContents() {
+		try {
+			if (contents != null) {
```
Comment (tivervac) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/util/AnnotatedTextToString.java:57]: You can move this outside of the the try

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractEditorDoubleClickTextSelectionTest.java [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	protected String getFileExtension() {
+		return fileExtensionProvider.getPrimaryFileExtension();
+	}
+
+	protected int getCursorPosition(CharSequence text) {
+		int cursorPosition = text.toString().indexOf(c());
+		if (cursorPosition == -1) {
+			StringConcatenation builder = new StringConcatenation();
```
[the remark this answers is not in the document]
Comment (nbhusare) [reply org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractEditorDoubleClickTextSelectionTest.java:96]: Sorry, what does the `---` mean?

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/util/AnnotatedTextToString.java [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+
+	public AnnotatedTextToString withMarkers(IMarker[] markers, String... attributes) {
+		try {
+			for (IMarker marker : markers) {
+				List<String> transformed = new ArrayList<String>(attributes.length);
+				for (String attribute : attributes) {
+					transformed.add(attribute + "=" + marker.getAttribute(attribute));
+				}
+
+				String comment = transformed.stream().collect(Collectors.joining(", "));
```
[the remark this answers is not in the document]
Comment (nbhusare) [reply org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/util/AnnotatedTextToString.java:108]: I'll use `String.join(", ", transformed)`. Tx for your suggestion. This is a better option than using the tedious `transformed.stream().collect(Collectors.joining(", "))`.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractHoverTest.java (hunk 10 above)
Comment (nbhusare) [reply org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractHoverTest.java]: PS - https://github.com/eclipse/xtext-eclipse/pull/1402/files#diff-c376f0d7106ef132b9b489fd869fddf1R80

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractHoverTest.java [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+ * @since 2.16
+ */
+@SuppressWarnings("restriction")
+public abstract class AbstractHoverTest extends AbstractEditorTest {
+
+	@Inject
+	protected IEObjectHover eObjectHover;
+
+	@Inject
+	protected SyncUtil syncUtil;
```
[the remark this answers is not in the document]
Comment (nbhusare) [reply org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractHoverTest.java]: Removed.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+			throw new RuntimeException(e);
+		}
+	}
+
+	protected void foldingRegionsArePresent(IXtextDocument document, CharSequence expected) {
+		assertEquals(expected.toString().trim(), insertFoldingRegionMarkers(document));
+	}
+
+	protected String insertFoldingRegionMarkers(IXtextDocument document) {
+		String lineSeparator = System.lineSeparator();
```
[the remark this answers is not in the document]
Comment (nbhusare) [reply org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java]: Added a new method `insertFoldingRegionMarkers(IXtextDOcument doc, String lineSeparator)`. Also checking if the document is empty before calling `document.getLineDelimiter(0)`. Please check.

**Comment by nbhusare** [comment]
@szarnekow I have addressed most of your review comments. Kindly have a look and resolve if you are ok with it.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractEditorDoubleClickTextSelectionTest.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (szarnekow) [reply org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractEditorDoubleClickTextSelectionTest.java:96]: I meant this as a delimiter to delimit the model text from the other part of the error message.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+		assertEquals(expected.toString().trim(), insertFoldingRegionMarkers(document));
+	}
+
+	protected String insertFoldingRegionMarkers(IXtextDocument document) {
+		if (document.getLength() == 0) { // The document is empty
+			return null;
+		}
+
+		try {
+			String lineSeparator = document.getLineDelimiter(0);
```
Comment (szarnekow) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java:100]: getLineDelimiter is document to return null if the line does not have a delimiter. I'd rather fix this here than in the overload below.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+			return insertFoldingRegionMarkers(document, lineSeparator);
+		} catch (BadLocationException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	protected String insertFoldingRegionMarkers(IXtextDocument document, String lineSeparator) {
+		if (lineSeparator == null) { // The document contains a single line
```
Comment (szarnekow) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java:108]: This method should assume that lineSeparator is not null.

**Comment by nbhusare** [comment]
@szarnekow I have updated the code. Please check.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractEditorDoubleClickTextSelectionTest.java [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+	}
+
+	protected String getFileExtension() {
+		return fileExtensionProvider.getPrimaryFileExtension();
+	}
+
+	protected int getCursorPosition(CharSequence text) {
+		int cursorPosition = text.toString().indexOf(c());
+		if (cursorPosition == -1) {
+			String message = "The input text\n %s \nmust contain the '%s' special symbol indicating the current cursor position!\n";
```
Comment (szarnekow) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractEditorDoubleClickTextSelectionTest.java:95]: Space after newline will be irritating. -> `\n%s\n` would be prefered

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractHoverTest.java [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+		XtextEditor editor = openInEditor(dslFile);
+		Object hoverInfo = ((ITextHoverExtension2) eObjectHover).getHoverInfo2(editor.getInternalSourceViewer(), hoverRegion);
+		return (BrowserInformationControlInput) hoverInfo;
+	}
+
+	protected void hoverPopupHasContent(BrowserInformationControlInput info, String expectedHoverContent) {
+		assertNotNull("No hover found!", info);
+
+		String actualHoverContent = info.getHtml();
+		String message = "Could not find the text '%s' in the hover popup:\n \t%s\n";
```
Comment (szarnekow) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractHoverTest.java:99]: `\n \t` -> please use `\n%s\n` instead

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+	protected String getFileName() {
+		return "folding";
+	}
+
+	protected String getFileExtension() {
+		return fileExtensionProvider.getPrimaryFileExtension();
+	}
+
+	protected String getContent(CharSequence text) {
+		return text.toString().replaceAll(Pattern.quote(b()), "").replaceAll(Pattern.quote(e()), "");
```
Comment (szarnekow) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractFoldingTest.java:77]: Can be simplified to `text.replace(b(), "").replace(e(), "")`

**Comment by nbhusare** [comment]
@szarnekow Please check.

**Comment by cdietrich** [comment]
thx @nbhusare

**Pull request merged**